test/test_utils_log.cpp

```diff
@@ -0,0 +1,380 @@
+                        } else {
+                            expect_fopen_count = 0;
+                        }
+                        if (logLevel.second != NO_CHANGE) {
+                            b.level = (utils_log_level_t)logLevel.second;
+                            if (logLevel.second <= LOG_INFO) {
+                                expect_fput_count = 1;
+                            } else {
+                                expect_fput_count = 0;
+                            }
+                        }
+                        helper_log_init(envVar.c_str());
+                        helper_checkConfig(&b, &loggerConfig);
+                    }
+                }
+            }
+        }
+    }
+}
+
+template <typename... Args> void helper_test_log(Args... args) {
+    fput_count = 0;
+    fflush_count = 0;
+    utils_log(args...);
+    EXPECT_EQ(fput_count, expect_fput_count);
+    EXPECT_EQ(fflush_count, expect_fflush_count);
+}
+
+static std::string helper_log_str(int l) {
+    switch (l) {
+    case LOG_DEBUG:
+        return "DEBUG";
+    case LOG_ERROR:
+        return "ERROR";
+    case LOG_INFO:
+        return "INFO ";
+    case LOG_WARNING:
+        return "WARN ";
+    default:
+        ASSERT(0);
+        return "";
+    }
+}
+
+TEST_F(test, log_levels) {
+    expected_stream = stderr;
+    for (int i = LOG_DEBUG; i <= LOG_ERROR; i++) {
+        for (int j = LOG_DEBUG; j <= LOG_ERROR; j++) {
+            loggerConfig = {0, 0, (utils_log_level_t)i, LOG_DEBUG, stderr};
+            if (i > j) {
+                expect_fput_count = 0;
+                expect_fflush_count = 0;
+                expected_message = "";
+            } else {
+                expect_fput_count = 1;
+                expect_fflush_count = 1;
+            }
+            expected_message = "[" + helper_log_str(j) + " UMF] example log\n";
+            helper_test_log((utils_log_level_t)j, "example log");
+        }
+    }
+}
+
+TEST_F(test, log_outputs) {
+    std::vector<FILE *> outs = {stdout, stderr, MOCK_FILE_PTR};
+    expect_fput_count = 1;
+    expect_fflush_count = 1;
+    expected_message = "[DEBUG UMF] example log\n";
+    for (auto o : outs) {
+        loggerConfig = {0, 0, LOG_DEBUG, LOG_DEBUG, o};
+        expected_stream = o;
+        helper_test_log(LOG_DEBUG, "example log");
+    }
+}
+
+TEST_F(test, flush_levels) {
+    expected_stream = stderr;
+    expect_fput_count = 1;
+    for (int i = LOG_DEBUG; i <= LOG_ERROR; i++) {
+        for (int j = LOG_DEBUG; j <= LOG_ERROR; j++) {
+            loggerConfig = {0, 0, LOG_DEBUG, (utils_log_level_t)i, stderr};
+            if (i > j) {
+                expect_fflush_count = 0;
+            } else {
+                expect_fflush_count = 1;
+            }
+            expected_message = "[" + helper_log_str(j) + " UMF] example log\n";
+            helper_test_log((utils_log_level_t)j, "example log");
+        }
+    }
+}
+
+TEST_F(test, long_log) {
```

Comment:
,can you run this test (and the next 3 tests) for various levels, not only DEBUG...?

Reply:
I kept this test "unit" we already tested that log level works - there is no need duplicate this tests and make those tests bigger.

Reply:
heh, the problem with unit tests is... that they are only unit tests :P We know both of these functionalities works independently, but we could have some higher-level tests as well. Nonetheless, it's ok for now, I guess.